### PR TITLE
Support `BERTModel` (first `encoder-only` embedding model)

### LINKED ISSUE #5179: [Feature]: BERT models for embeddings
Now with the introduction of embeddings: https://github.com/vllm-project/vllm/pull/3734, are there plans on the roadmap to support BERT models?

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
SUMMARY:
* built on top of https://github.com/vllm-project/vllm/pull/5447 (co-author: [laishzh](https://github.com/laishzh))
* introduces Encoder-Only models (BERT - tested with https://huggingface.co/BAAI/bge-base-en-v1.5 to match sentence-transformers implementation)
* introduces Encoder-Only attention for XFORMERS backend. 

Note: this PR requires setting the `VLLM_ATTENTION_BACKEND=XFORMERS` variable to run. We throw a loud error if it is not set.

TODO:
* add end-to-end testing with MTEB
* test tp>1

FOLLOW UPS:
* refactor to extract some of logic out of XFORMERS. The logic is very complex right now.
* support other attention backend

FIX #5179

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by DarkLight1337** [comment]
I see both #5179 and this PR are being worked on. Which one should we review?

**Comment by robertgshaw2-neuralmagic** [comment]
We need a follow up to clean up the `XFormersAttention`

File: vllm/model_executor/layers/pooler.py [hunk 1]
```diff
@@ -23,12 +24,13 @@ class Pooler(nn.Module):
     3. Returns structured results as `PoolerOutput`.
 
     Attributes:
-        pooling_type: The type of pooling to use (LAST, AVERAGE, MAX).
+        pooling_type: The type of pooling to use (LAST, MAX, CLS).
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:27]:
```suggestion
        pooling_type: The type of pooling to use (LAST, ALL, CLS).
```

Right now, we don't have `MAX`...

File: tests/models/embedding/language/test_embedding.py [hunk 2]
```diff
@@ -2,15 +2,20 @@
...
 import torch
 import torch.nn.functional as F
 
 MODELS = [
+    "BAAI/bge-base-en-v1.5",
     "intfloat/e5-mistral-7b-instruct",
     "BAAI/bge-multilingual-gemma2",
 ]
 
+ENCODER_MODELS = ["BAAI/bge-base-en-v1.5"]
```
Comment (DarkLight1337) [inline tests/models/embedding/language/test_embedding.py:17]:
```suggestion
ENCODER_ONLY_MODELS = ["BAAI/bge-base-en-v1.5"]
```

To distinguish it from encoder-decoder models, similar to what you've done in `attention` module.

**Review by DarkLight1337 — COMMENTED** [review]
I can't say to the implementation of the model (don't have time to compare it against HF), but nothing really jumps out at me from a brief look. If the tests pass then it's probably fine.

File: tests/models/embedding/language/test_embedding.py [hunk 3]
```diff
@@ -49,3 +59,9 @@ def test_models(
     assert torch.all((all_similarities <= 1.0 + tolerance)
                      & (all_similarities >= 1.0 - tolerance)
                      ), f"Not all values are within {tolerance} of 1.0"
```
Comment (DarkLight1337) [inline tests/models/embedding/language/test_embedding.py:61]:
```suggestion
    assert torch.all(all_similarities >= 1.0 - tolerance), f"Not all values are within {tolerance} of 1.0"
```

The first check is unnecessary since similarity values shouldn't exceed 1.

**Comment by robertgshaw2-neuralmagic** [comment]
> I can't say to the implementation of the model (don't have time to compare it against HF), but nothing really jumps out at me from a brief look. If the tests pass then it's probably fine.

I have a test that compares against the hf implementation (via sentence-transformers)

File: vllm/model_executor/models/bert.py [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+                 config: BertConfig,
+                 cache_config: Optional[CacheConfig] = None,
+                 quant_config: Optional[QuantizationConfig] = None,
+                 prefix: str = ""):
+        super().__init__()
+        self.embeddings = BertEmbedding(config)
+        self.encoder = BertEncoder(config,
+                                   cache_config,
+                                   quant_config,
+                                   prefix=f"{prefix}.encoder")
```
Comment (pooyadavoodi) [inline vllm/model_executor/models/bert.py:322]: I wonder if we should set this to `prefix=f"{prefix}.encoder_only")`

File: vllm/model_executor/models/bert.py [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+        intermediate_tensors: Optional[IntermediateTensors] = None,
+        inputs_embeds: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        if inputs_embeds is not None:
+            hidden_states = inputs_embeds
+        else:
+            hidden_states = self.embeddings(input_ids=input_ids,
+                                            position_ids=position_ids)
+
+        return self.encoder(hidden_states, kv_caches, attn_metadata)
```
Comment (pooyadavoodi) [inline vllm/model_executor/models/bert.py:339]: My understanding is that kv_caches would be an empty list here. Do we validate that some where?

File: vllm/model_executor/models/bert.py [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+        kv_caches: List[torch.Tensor],
+        attn_metadata: AttentionMetadata,
+    ) -> torch.Tensor:
+        for i in range(len(self.layer)):
+            layer = self.layer[i]
+            hidden_states = layer(hidden_states, kv_caches[i], attn_metadata)
+        return hidden_states
+
+
+class BertLayer(nn.Module):
```
Comment (pooyadavoodi) [inline vllm/model_executor/models/bert.py:99]: I think the coding style is to put the definition of dependencies above their usage, like `BertLayer` before `BertEncoder`?

File: vllm/model_executor/models/bert.py [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+        hidden_size: int,
+        num_attention_heads: int,
+        layer_norm_eps: float,
+        cache_config: Optional[CacheConfig] = None,
+        quant_config: Optional[QuantizationConfig] = None,
+        prefix: str = "",
+    ):
+        super().__init__()
+
+        self.self = BertSelfAttention(hidden_size=hidden_size,
```
Comment (pooyadavoodi) [inline vllm/model_executor/models/bert.py:154]: nit: `self.self` --> `self.self_attn`

File: vllm/model_executor/models/bert.py [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+                 layer_norm_eps: float,
+                 quant_config: Optional[QuantizationConfig] = None,
+                 prefix: str = ""):
+        super().__init__()
+        self.dense = RowParallelLinear(input_size=hidden_size,
+                                       output_size=hidden_size,
+                                       bias=True,
+                                       quant_config=quant_config,
+                                       prefix=f"{prefix}.dense")
+        self.LayerNorm = nn.LayerNorm(hidden_size, eps=layer_norm_eps)
```
Comment (pooyadavoodi) [inline vllm/model_executor/models/bert.py:254]: nit: `self.LayerNorm` --> `self.layer_norm`

**Review by pooyadavoodi — CHANGES_REQUESTED** [review]

File: vllm/model_executor/models/bert.py (hunk 4 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/models/bert.py:322]: The prefix needs to match the state dict exactly.

File: vllm/model_executor/models/bert.py (hunk 5 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/models/bert.py:339]: I believe these should be empty. Will add an assert.

File: vllm/model_executor/models/bert.py (hunk 6 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/models/bert.py:99]: Good idea

File: vllm/model_executor/models/bert.py (hunk 7 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/models/bert.py:154]: this needs to match the name in the huggingface state dict exactly for weight loading to work, so unfortunately this cannot change

File: vllm/model_executor/models/bert.py (hunk 8 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/models/bert.py:254]: this needs to match the name in the huggingface state dict exactly for weight loading to work, so unfortunately this cannot change

File: vllm/model_executor/models/bert.py [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+
+    def __init__(
+        self,
+        config: BertConfig,
+        cache_config: Optional[CacheConfig] = None,
+        quant_config: Optional[QuantizationConfig] = None,
+    ) -> None:
+        super().__init__()
+        self.model = BertModel(config, cache_config, quant_config)
+        self._pooler = Pooler(pooling_type=PoolingType.CLS, normalize=True)
```
Comment (maxdebayser) [inline vllm/model_executor/models/bert.py:393]: In the `transformers` code for this model I don't see normalization being applied: https://github.com/huggingface/transformers/blob/d00f1ca860f19f4c0962882e56044bb6ef7b5626/src/transformers/models/bert/modeling_bert.py#L743 
How do we know if it has to be applied or not?

**Review by maxdebayser — COMMENTED** [review]
I've tested it with BAAI/bge-base-en-v1.5 and get results close to the MTEB benchmark.

File: vllm/model_executor/models/bert.py [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        position_ids: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        input_shape = input_ids.size()
+
+        # Input embeddings.
+        inputs_embeds = self.word_embeddings(input_ids)
+
```
Comment (maxdebayser) [inline vllm/model_executor/models/bert.py:56]: In `transformers`, if  `position_ids` is None, `self.position_ids` is used. Should we do the same?

https://github.com/huggingface/transformers/blob/d00f1ca860f19f4c0962882e56044bb6ef7b5626/src/transformers/models/bert/modeling_bert.py#L196

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/models/bert.py:56]: Position Ids will never be None for us

**Review by maxdebayser — APPROVED** [review]

File: tests/models/embedding/language/test_embedding.py [hunk 11]
```diff
@@ -43,3 +53,9 @@ def test_models(
         name_1="vllm",
         tol=1e-2,
     )
+
+    if "VLLM_ATTENTION_BACKEND" in os.environ:
+        if prior_attn_backend_env is None:
+            del os.environ["VLLM_ATTENTION_BACKEND"]
+        else:
+            os.environ["VLLM_ATTENTION_BACKEND"] = prior_attn_backend_env
```
Comment (DarkLight1337) [inline tests/models/embedding/language/test_embedding.py:61]: To avoid missing this when the test fails (which may be relevant when multiple tests are run in sequence), I recommend placing this in a pytest fixture.

**Review by pooyadavoodi — APPROVED** [review]

File: tests/models/embedding/language/test_embedding.py [hunk 12]
```diff
@@ -2,25 +2,49 @@
...
+
+@pytest.fixture
+def encoder_env_var_guard():
+    prior_attn_backend_env = os.getenv("VLLM_ATTENTION_BACKEND", None)
+    os.environ["VLLM_ATTENTION_BACKEND"] = "XFORMERS"
+    yield None
+    if prior_attn_backend_env is None:
+        del os.environ["VLLM_ATTENTION_BACKEND"]
+    else:
+        os.environ["VLLM_ATTENTION_BACKEND"] = prior_attn_backend_env
```
Comment (DarkLight1337) [inline tests/models/embedding/language/test_embedding.py:20]: I guess you can remove this now then.

**Review by DarkLight1337 — APPROVED** [review]
Overall LGTM - remember to add this model to the Supported Models page before you merge.

**Comment by DarkLight1337** [comment]
Is it intentional to not include this on the docs yet?

**Comment by robertgshaw2-neuralmagic** [comment]
> Is it intentional to not include this on the docs yet?

Yes I think we should get #9388 in before we advertise this
